### PR TITLE
StreamEffect optimizations

### PR DESCRIPTION
**Pull request by vasilmkd**
I'm aware that this is not the nicest looking Scala code (manually instantiating `Function0`) but I believe it is worth it.

1. I have removed a whole layer of `Managed` instances on _each_ combinator.
2. I have removed all captured variables in closures (thus avoiding boxing), by making them local class fields (something that the JVM should be absolutely best at).

@jdegoes @iravid 

EDIT: See benchmark results below. I'm just about ready to admit defeat.

### CODE REVIEW

**Comment by vasilmkd** [comment]
I'll work on some benchmarks.

**Comment by vasilmkd** [comment]
```scala
  @Benchmark
  def scalazFilterMapSum = {
    val stream = Stream.iterate(1L)(_ + 1L).take(10000000).filter(_ % 2 == 0).map(_ * 2).filter(_ % 2 == 0).map(_ * 2)
    val sink = Sink.foldLeft(0L)((sum, n: Long) => sum + n)
    val result = stream.run(sink)
    unsafeRun(result)
  }
```

[sum-bench-master.txt](https://github.com/zio/zio/files/3924327/sum-bench-master.txt)
[sum-bench-opt.txt](https://github.com/zio/zio/files/3924328/sum-bench-opt.txt)

**Comment by vasilmkd** [comment]
Definitely not an expert, but I suspect that by instantiating real objects instead of lambdas (as Scala 2.12 source code is translated to jvm8 bytecode), we are foregoing `invokedynamic` for `invokeinterface`.

**Comment by iravid** [comment]
Hmm it appears to be slower 🧐

**Comment by vasilmkd** [comment]
Yeah, it appears so. Worth a try. Feel free to close the PR.

**Comment by iravid** [comment]
Would you like to investigate why it is slower? Or not interested?

**Comment by vasilmkd** [comment]
I think it's worth a look. Can you give me a few pointers?

**Comment by iravid** [comment]
Try checking what those `Function0` objects are compiling into. Ideally the `var`s should be simple fields with no getter/setters. Also compare the bytecode to master to see if it's actually simplifying it.

**Comment by vasilmkd** [comment]
I did that before. There was a getter. Can I maybe force direct access with a `public var`?

See some of my previous comments. I think the biggest change is losing `invokedynamic` lambdas, and instead doing an `invokeinterface` on a `Function0` object.

**Comment by iravid** [comment]
`private[this]` might force direct field access

**Comment by vasilmkd** [comment]
I'll try that. Other than that, there was definitely no boxing through `ObjectRef` and friends.

**Comment by vasilmkd** [comment]
You're right, `private[this]` does force direct access, but it's still slower.

Master:
[sum-bench-master-private-this.txt](https://github.com/zio/zio/files/3925986/sum-bench-master-private-this.txt)
---
This PR:
[sum-bench-pr-private-this.txt](https://github.com/zio/zio/files/3925987/sum-bench-pr-private-this.txt)

**Comment by vasilmkd** [comment]
@iravid Here are the .class files if you'd like to take a look. They are too long to paste here.

[classes.zip](https://github.com/zio/zio/files/3926235/classes.zip)

**Comment by vasilmkd** [comment]
@iravid Browsing the bytecode, I noticed that `ZSink` methods are being called with `invokeinterface`. Since we have migrated `ZStream` to be a class, do you think it's time to migrate `ZSink` to become an `abstract class` to unlock some more optimizations from the JVM?

File: streams/shared/src/main/scala/zio/stream/StreamEffect.scala [hunk 1]
```diff
@@ -37,68 +37,61 @@ private[stream] class StreamEffect[-R, +E, +A](val processEffect: ZManaged[R, No
...
         }
       }
     }
 
   override def collectWhile[B](pred: PartialFunction[A, B]): StreamEffect[R, E, B] =
     StreamEffect {
-      self.processEffect.flatMap { thunk =>
-        Managed.effectTotal {
+      self.processEffect.map { thunk =>
+        new Function0[B] {
```
Comment (regiskuckaertz) [inline streams/shared/src/main/scala/zio/stream/StreamEffect.scala]: Can you memoise the object instead of recreating it each time?

Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/StreamEffect.scala]: That one, maybe, but what I really set out to fix with this PR was to see if the variable capture boxing penalty could be reduced, and for those, I don't think that memoization of the function is an option, given the mutable state.

Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/StreamEffect.scala]: Ah yeah, I see. So you defining a `Function2` and then currying it wouldn't help I guess 😞

Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/StreamEffect.scala]: I thought about that too haha. The thing is, as soon as you introduce a ref type to a function, regardless of whether it is an argument or result, the function is not specialized anymore.

**Comment by vasilmkd** [comment]
This was a fun experiment.

**Pull request closed without merging**